File: app/src/main/java/com/zoneanchor/ZonedAlarm.java
package com.zoneanchor;

public final class ZonedAlarm {
    public static final int FIRST_ID = 400;
    public static final int DEFAULT_ID = FIRST_ID;
    public static final String DEFAULT_ZONE_ID = "America/New_York";
    public static final int DEFAULT_HOUR = 16;
    public static final int DEFAULT_MINUTE = 0;
    public static final int ALL_DAYS = 0b1111111;
    public static final String SOUND_DEFAULT = "default";
    public static final String SOUND_SILENT = "silent";

    public final int id;
    public final String label;
    public final String zoneId;
    public final int hour;
    public final int minute;
    public final int daysOfWeekMask;
    public final String soundMode;
    public final boolean vibrate;
    public final boolean enabled;
    public final long nextTriggerAtMillis;

    public ZonedAlarm(
            int id,
            String label,
            String zoneId,
            int hour,
            int minute,
            int daysOfWeekMask,
            String soundMode,
            boolean vibrate,
            boolean enabled,
            long nextTriggerAtMillis
    ) {
        this.id = id;
        this.label = cleanLabel(label);
        this.zoneId = zoneId;
        this.hour = hour;
        this.minute = minute;
        this.daysOfWeekMask = daysOfWeekMask == 0 ? ALL_DAYS : daysOfWeekMask;
        this.soundMode = SOUND_SILENT.equals(soundMode) ? SOUND_SILENT : SOUND_DEFAULT;
        this.vibrate = vibrate;
        this.enabled = enabled;
        this.nextTriggerAtMillis = nextTriggerAtMillis;
    }

    public static ZonedAlarm defaultAlarm() {
        return new ZonedAlarm(
                DEFAULT_ID,
                "",
                DEFAULT_ZONE_ID,
                DEFAULT_HOUR,
                DEFAULT_MINUTE,
                ALL_DAYS,
                SOUND_DEFAULT,
                true,
                false,
                0L
        );
    }

    public ZonedAlarm enabledWithNextTrigger(long nextTriggerAtMillis) {
        return new ZonedAlarm(
                id,
                label,
                zoneId,
                hour,
                minute,
                daysOfWeekMask,
                soundMode,
                vibrate,
                true,
                nextTriggerAtMillis
        );
    }

    public ZonedAlarm disabled() {
        return new ZonedAlarm(
                id,
                label,
                zoneId,
                hour,
                minute,
                daysOfWeekMask,
                soundMode,
                vibrate,
                false,
                0L
        );
    }

    public String displayName() {
        return label.isEmpty() ? zoneId : label;
    }

    private static String cleanLabel(String value) {
        return value == null ? "" : value.trim();
    }
}

File: app/src/main/java/com/zoneanchor/model/ZonedAlarm.kt
package com.zoneanchor.model

import java.time.ZoneId

data class ZonedAlarm(
    val id: Int = DEFAULT_ID,
    val label: String = "",
    val zoneId: String = DEFAULT_ZONE_ID,
    val hour: Int = DEFAULT_HOUR,
    val minute: Int = DEFAULT_MINUTE,
    val daysOfWeekMask: Int = ALL_DAYS,
    val soundMode: String = SOUND_DEFAULT,
    val vibrate: Boolean = true,
    val enabled: Boolean = false,
    val nextTriggerAtMillis: Long = 0L,
) {
    val displayName: String get() = label.ifBlank { zoneId }
    val normalizedDaysMask: Int get() = normalizeDays(daysOfWeekMask)

    fun enabledWithNextTrigger(nextTrigger: Long) = copy(
        label = cleanLabel(label),
        zoneId = validZoneId(zoneId),
        daysOfWeekMask = normalizedDaysMask,
        soundMode = validSoundMode(soundMode),
        enabled = true,
        nextTriggerAtMillis = nextTrigger,
    )

    fun disabled() = copy(enabled = false, nextTriggerAtMillis = 0L)

    companion object {
        const val FIRST_ID = 400
        const val DEFAULT_ID = FIRST_ID
        const val DEFAULT_ZONE_ID = "America/New_York"
        const val DEFAULT_HOUR = 16
        const val DEFAULT_MINUTE = 0
        const val ALL_DAYS = 0b1111111
        const val SOUND_DEFAULT = "default"
        const val SOUND_SILENT = "silent"

        fun cleanLabel(value: String?) = value?.trim().orEmpty()

        fun normalizeDays(mask: Int): Int {
            val normalized = mask and ALL_DAYS
            return if (normalized == 0) ALL_DAYS else normalized
        }

        fun validSoundMode(value: String?) =
            if (value == SOUND_SILENT) SOUND_SILENT else SOUND_DEFAULT

        fun validZoneId(value: String?): String {
            return try {
                ZoneId.of(value ?: DEFAULT_ZONE_ID).id
            } catch (_: Exception) {
                DEFAULT_ZONE_ID
            }
        }
    }
}
